src/Http/Controllers/Cp/ProductController.php

```diff
@@ -204,49 +202,41 @@ public function update(ProductUpdateRequest $request, Product $product)
                 ]);
 
                 $requestAttributes = collect($variant['variant_attributes'])
-                    ->each(function ($attribute) use ($item) {
+                    ->map(function ($attribute) use ($item) {
                         if ($attribute['key'] === null) {
-                            return;
+                            return $attribute;
                         }
 
-                        $item->attributes()->updateOrCreate([
+                        $attributeRecord = $item->attributes()->updateOrCreate([
                             'uuid' => $attribute['uuid'],
                         ], [
                             'key' => $attribute['key'],
                             'value' => $attribute['value'],
                         ]);
+
+                        $attribute['uuid'] = $attributeRecord->uuid;
+
+                        return $attribute;
                     });
 
                 $item->attributes
                     ->filter(function ($attribute) use ($requestAttributes) {
                         return ! $requestAttributes
                             ->contains(function ($requestAttribute) use ($attribute) {
-                                return $attribute->key === $requestAttribute['key'] && $attribute->value === $requestAttribute['value'];
+                                return $attribute->uuid === $requestAttribute['uuid'];
                             });
                     })
                     ->each->delete();
-            });
 
-        $requestAttributes = collect($request->product_attributes)
-            ->each(function ($attribute) use ($product) {
-                if ($attribute['key'] === null) {
-                    return;
-                }
+                $variant['uuid'] = $item->uuid;
 
-                $product->attributes()->updateOrCreate([
-                    'uuid' => $attribute['uuid'],
-                ], [
-                    'key' => $attribute['key'],
-                    'value' => $attribute['value'],
-                ]);
+                return $variant;
             });
 
-        $product->attributes
-            ->filter(function ($attribute) use ($requestAttributes) {
-                return ! $requestAttributes
-                    ->contains(function ($requestAttribute) use ($attribute) {
-                        return $attribute->key === $requestAttribute['key'] && $attribute->value === $requestAttribute['value'];
-                    });
+        $product->variants
+            ->filter(function ($variant) use ($requestVariants) {
```

Comment:
where is that getting `$requestVariants` from?

Reply:
> where is that getting `$requestVariants` from?

Line 189